### ISSUE TITLE
How to use new snapshotting?

### ISSUE THREAD

**Issue by xksteven** [rbgirshick/fast-rcnn#35]
Do I call ./tools/train_net.py with --snapshot=snapshot.caffemodel ?

I'm asking because in the README it has this :
"We disable standard caffe solver snapshotting and implement our own snapshot"

Thanks

**Comment by WilsonWangTHU** [comment]
It's in the ./lib/fast-rcnn/config.py

**Comment by xksteven** [comment]
In that file I can change the time between snapshots and snapshot infix but nothing on using the snapshot during training.

Would I just change in the solver.prototxt the snapshot number to reference the current snapshot?

**Comment by WilsonWangTHU** [comment]
@xksteven  I guess you would like to do the validation during the training?
I am not sure whether that's supported by the current fast-rcnn edition, as all the forward job is started from the python part and I don't think we have a testing function during the training for now. 
I am afraid in that way you might need to revise the code yourself.

**Comment by xksteven** [comment]
@WilsonWangTHU 
You know when using caffe you can provide the snapshot option such as -snapshot=model_iter_xxx.solverstate to restart the training from that point? Normally in caffe the solverstate and the caffemodel saved as model_iter_xxx.caffemodel are both in the same directory but with fast-rcnn I only see the caffemodel saved in the output/default/imdb_trainval.  I'd like to be able to restart the training using those weights stored there.  

I'm running it on a cluster with a certain time limit and it will kill my process at certain time intervals. I just want to be able to restart the training from that snapshot.  
thanks for your help.

**Comment by kyuusaku** [comment]
I have the same problem.

**Comment by kyuusaku** [comment]
How to restart the training from a snapshot? Can anyone provide some tips? Thanks.

**Comment by IdiosyncraticDragon** [comment]
@kyuusaku @xksteven I have met the same problem, do you guys get some effective solutions?Thanks

**Comment by lynetcha** [comment]
Make the following modifications and you will be able to use the --snapshot argument

**In tools/train_net.py**
<pre>

    def parse_args():
        """
        Parse input arguments
        """
        parser = argparse.ArgumentParser(description='Train a Fast R-CNN network')
        parser.add_argument('--gpu', dest='gpu_id',
                            help='GPU device id to use [0]',
                            default=0, type=int)
        parser.add_argument('--solver', dest='solver',
                            help='solver prototxt',
                            default=None, type=str)
        parser.add_argument('--iters', dest='max_iters',
                            help='number of iterations to train',
                            default=40000, type=int)
        parser.add_argument('--weights', dest='pretrained_model',
                            help='initialize with pretrained model weights',
                            default=None, type=str)
        <b>parser.add_argument('--snapshot', dest='previous_state',
                            help='initialize with previous state',
                            default=None, type=str) </b>
        parser.add_argument('--cfg', dest='cfg_file',
                            help='optional config file',
                            default=None, type=str)
        parser.add_argument('--imdb', dest='imdb_name',
                            help='dataset to train on',
                            default='voc_2007_trainval', type=str)
        parser.add_argument('--rand', dest='randomize',
                            help='randomize (do not use a fixed seed)',
                            action='store_true')
        parser.add_argument('--set', dest='set_cfgs',
                            help='set config keys', default=None,
                            nargs=argparse.REMAINDER)

</pre>


 **In lib/fast_rcnn/train.py**
<pre>
        class SolverWrapper(object):
            """A simple wrapper around Caffe's solver.
            This wrapper gives us control over he snapshotting process, which we
            use to unnormalize the learned bounding-box regression weights.
            """

            def __init__(self, solver_prototxt, roidb, output_dir,
                         pretrained_model=None, <b>previous_state=None</b>):
                """Initialize the SolverWrapper."""
                self.output_dir = output_dir

                print 'Computing bounding-box regression targets...'
                self.bbox_means, self.bbox_stds = \
                        rdl_roidb.add_bbox_regression_targets(roidb)
                print 'done'

                self.solver = caffe.SGDSolver(solver_prototxt)
                if pretrained_model is not None:
                    print ('Loading pretrained model '
                           'weights from {:s}').format(pretrained_model)
                    self.solver.net.copy_from(pretrained_model)
                <b> elif previous_state is not None:
                    print ('Restoring State from '
                              ' from {:s}').format(previous_state)
                    self.solver.restore(previous_state)</b>

                self.solver_param = caffe_pb2.SolverParameter()
                with open(solver_prototxt, 'rt') as f:
                    pb2.text_format.Merge(f.read(), self.solver_param)

                self.solver.net.layers[0].set_roidb(roidb)
.
.
.
def train_net(solver_prototxt, roidb, output_dir,
              pretrained_model=None, max_iters=40000,<b>previous_state=None</b>):
    """Train a Fast R-CNN network."""
    sw = SolverWrapper(solver_prototxt, roidb, output_dir,
                       pretrained_model=pretrained_model,<b>previous_state=previous_state</b>)

    print 'Solving...'
    sw.train_model(max_iters)
    print 'done solving'

</pre>

**Comment by chrert** [comment]
Thanks for the code but how to save the solverstate during fast r-cnn training? It looks like the method Solver::SnapshotSolverState isn't exported to pycaffe...

**Comment by lynetcha** [comment]
Did you change "snapshot: 0"  to "snapshot: 10000" in your solver.prototxt?  That allows you to save the state at iteration 10000 for example.

**Comment by chrert** [comment]
Ah, thanks! Didn't think of that...

**Comment by enkhee0906** [comment]
I did a small experiment using 8 layer CaffeNet on VOC2007 set in order to check whether resuming training works well in Fast RCNN.  I saved weight and solver state using snapshot at every 10,000 iteration and stopped training. Then resumed training by restoring solver state. I iteratively did this 4 times, so total number of iterations is 40,000, the same number of iterations in Fast RCNN. So it's supposed to give me similar result as wihout resuming training. But detection performance is 55.2% which means ~2% different from without resuming training. In the paper paper, detection performance is 57.1 on VOC2007 set when using CaffeNet. I don't know what is the problem.

**Comment by smichalowski** [comment]
@lynetcha, one more modification:

**In tools/train_net.py**

```
    output_dir = get_output_dir(imdb)
    print 'Output will be saved to `{:s}`'.format(output_dir)

    train_net(args.solver, roidb, output_dir,
              pretrained_model=args.pretrained_model,
              max_iters=args.max_iters, **previous_state=args.previous_state**)
```

also remember to omit --weights param

**Comment by SaiAdityaG** [comment]
I was able to resume training in end-to-end method. But how to do the same in 'alternate' training method too.

**Comment by po0ya** [comment]
Thanks all for the info. I think a PR would be appropriate for this. The final thing is that fast-rcnn re-scales the weights and the bias of the last layer of the regressor. The following quick fix change will take care of this by saving an extra file for the last layer weights. There must be better ways though.

    def snapshot(self):
        """Take a snapshot of the network after unnormalizing the learned
        bounding-box regression weights. This enables easy use at test-time.
        """
        net = self.solver.net

        if cfg.TRAIN.BBOX_REG:
            # save the original values
            orig_0 = net.params['bbox_pred'][0].data.copy()
            orig_1 = net.params['bbox_pred'][1].data.copy()
           

            # scale and shift with bbox reg unnormalization; then save snapshot
            net.params[bbox_pred][0].data[...] = \
                    (net.params[bbox_pred][0].data *
                     self.bbox_stds[:, np.newaxis])
            net.params[bbox_pred][1].data[...] = \
                    (net.params[bbox_pred][1].data *
                     self.bbox_stds + self.bbox_means)

        if not os.path.exists(self.output_dir):
            os.makedirs(self.output_dir)

        infix = ('_' + cfg.TRAIN.SNAPSHOT_INFIX
                 if cfg.TRAIN.SNAPSHOT_INFIX != '' else '')
        filename = (self.solver_param.snapshot_prefix + infix +
                    '_iter_{:d}'.format(self.solver.iter) + '.caffemodel')
        **filename_meta = (self.solver_param.snapshot_prefix + infix +
                    '_iter_{:d}'.format(self.solver.iter) + '.meta')**

        params = {}
        params['bbox'] = [orig_0,orig_1]

        filename = os.path.join(self.output_dir, filename)
        filename_meta = os.path.join(self.output_dir,filename_meta)
        net.save(str(filename))
        **with open(filename_meta,'w') as f:
            cPickle.dump(params,f)**

        print 'Wrote snapshot to: {:s}'.format(filename)
        if cfg.TRAIN.BBOX_REG:
            # restore net to original state
            net.params['bbox_pred'][0].data[...] = orig_0
            net.params['bbox_pred'][1].data[...] = orig_1
 
and then in the constructor of SolverWrapper:

        meta_file = pretrained_model.replace('caffemodel', 'meta')
        if os.path.exists(meta_file):
            with open(meta_file,'r') as f:
                params = cPickle.load(f)
                [orig_0,orig_1] = params['bbox']
                net.params[bbox_pred][0].data[...] = orig_0
                net.params[bbox_pred][1].data[...] = orig_1

**Comment by twmht** [comment]
hi @po0ya 

what if I don't save the extra file for the last layer weights? would be bad mAP when retraining?

**Comment by po0ya** [comment]
Hello @twmht 

Basically it'll mess up the whole network if you want to continue training. The network is trained to work for zero mean and unit variance bboxes. For test time convenience, the weights and bias of the last layer is scaled by the std and shifted by the mean. If it has not been done, the prediction should've been scaled and shifted manually. It's for convenience in testing time, but the weights are not the ones that were learned by backprop, so retraining with these weights would be meaningless for the network.

**Comment by ds2268** [comment]
@po0ya but aren't the weights (*.caffemodel) that are saved by the default solver already normalized (because they were never unnormalizied, because the caffemodel was not saved using provided snapshot functionality). So I guess the produced *.solverstate is linked to the *.caffemodel model that was not produced by the faster rcnn snapshot function. Using resuming functionality you get 2 versions of caffemodel, the one provided by the default solver snapshot and the one provided by the snapshot function in faster r-cnn that the weights are unnormalized before saving.

**Comment by misssprite** [comment]
Net params in snapshot function in SolverWrapper is first unnormalized,  saved and restored with normalized version. So the param version is up to when the snapshot in Caffe is called. 

I didn't dig the code of Caffe, but I think disabling snapshot in `solver.prototxt` and manually calling `solver.snapshot()` will be better to control exactly which version is snapshotted.

Actually, I look into the log and found that the Caffe snapshot is called **before** snapshot in `SolverWrapper`.  `diff` the params file shows that Caffe snapshot indeed save a different (normalized) version than `SolverWrapper`. Manually invocation of `solver.snapshop` obtained a identical `.caffemodel`.

So we can resume the `.solverstate` safely without unnormalizing the parameters with Caffe snapshot. But this produces two version of '.caffemodel's. It's up to you to snapshot which version of parameters.
